File: frontend/src/app/core/fail2ban.models.ts
/** A fail2ban jail with its counters and currently banned IPs. */
export interface Fail2banJail {
  name: string;
  currently_failed: number;
  total_failed: number;
  currently_banned: number;
  total_banned: number;
  /** IPs currently banned in this jail. */
  banned_ips: string[];
  /** Log files watched by the jail's filter. */
  file_list: string[];
}

/** Aggregated fail2ban status returned by GET /api/fail2ban/status. */
export interface Fail2banStatus {
  jails: Fail2banJail[];
  /** False when ENABLE_FAIL2BAN=0: no daemon runs, so no jail exists. */
  fail2ban_enabled: boolean;
}

/** A single banned IP and the jail it is banned in. */
export interface BannedIp {
  ip: string;
  jail: string;
}

export interface BanRequest {
  ip: string;
}

/** Raw command output returned after a ban/unban action. */
export interface Fail2banActionResult {
  output: string;
}

/** Trailing lines of the fail2ban log file. */
export interface Fail2banLog {
  lines: string[];
}

/** The ban policy stored in fail2ban-jail.cf under [DEFAULT]. */
export interface Fail2banPolicy {
  /** Seconds a banned IP stays banned. */
  bantime: number;
  /** Seconds over which failures are counted towards maxretry. */
  findtime: number;
  /** Failures within findtime before an IP is banned. */
  maxretry: number;
  /** False when no policy file exists yet: the values are docker-mailserver's defaults. */
  configured: boolean;
  /** The mailserver only reads the policy when it starts. */
  restart_required: boolean;
}

export interface Fail2banPolicyUpdateRequest {
  bantime: number;
  findtime: number;
  maxretry: number;
}

/**
 * The raw contents of fail2ban-fail2ban.cf: the daemon's own options — log level,
 * database retention — rather than any jail.
 */
export interface Fail2banConfig {
  content: string;
  /** Copied to /etc/fail2ban/fail2ban.local at startup: an edit needs a restart. */
  restart_required: boolean;
}

export interface Fail2banConfigUpdateRequest {
  content: string;
}

File: backend/app/models/fail2ban_models.py
"""API schemas for fail2ban management inside docker-mailserver.

Fail2ban runs inside the mailserver container and exposes no shared config
file, so — unlike mailboxes or relays — it is driven with ``docker exec``
(``fail2ban-client`` for structured reads, ``setup fail2ban`` for actions).
See :mod:`app.services.fail2ban_service`. These schemas describe the
request/response shapes only.
"""

from pydantic import BaseModel, Field


class Fail2banJail(BaseModel):
    """A single fail2ban jail with its counters and currently banned IPs."""

    name: str
    currently_failed: int = 0
    total_failed: int = 0
    currently_banned: int = 0
    total_banned: int = 0
    # IPs currently banned in this jail.
    banned_ips: list[str] = Field(default_factory=list)
    # Log files watched by the jail's filter.
    file_list: list[str] = Field(default_factory=list)


class Fail2banStatus(BaseModel):
    """Aggregated fail2ban status: every configured jail."""

    jails: list[Fail2banJail] = Field(default_factory=list)


class BannedIp(BaseModel):
    """A single banned IP and the jail it is banned in."""

    ip: str
    jail: str


class BanRequest(BaseModel):
    """Request schema to ban an IP address."""

    ip: str = Field(min_length=1, max_length=45)


class Fail2banActionResult(BaseModel):
    """Result of a ban/unban action: the raw command output, for feedback."""

    output: str = ""


class Fail2banLog(BaseModel):
    """Trailing lines of the fail2ban log file."""

    lines: list[str] = Field(default_factory=list)


class Fail2banPolicy(BaseModel):
    """The ban policy stored in ``fail2ban-jail.cf`` under ``[DEFAULT]``.

    docker-mailserver copies that file to ``/etc/fail2ban/jail.d/user-jail.local``
    when it starts, so a changed policy only takes effect after a restart.
    """

    # Seconds a banned IP stays banned.
    bantime: int = Field(ge=1)
    # Seconds over which failures are counted towards ``maxretry``.
    findtime: int = Field(ge=1)
    # Failures within ``findtime`` before an IP is banned.
    maxretry: int = Field(ge=1)
    # False when no ``fail2ban-jail.cf`` exists yet: the values above are then
    # the defaults docker-mailserver ships with, not a stored policy.
    configured: bool = False
    restart_required: bool = True


class Fail2banPolicyUpdate(BaseModel):
    """Request schema replacing the fail2ban ``[DEFAULT]`` ban policy."""

    # One week is the docker-mailserver default; cap at a year to stay sane.
    bantime: int = Field(ge=60, le=31_536_000)
    findtime: int = Field(ge=60, le=31_536_000)
    maxretry: int = Field(ge=1, le=100)


class Fail2banConfig(BaseModel):
    """The raw contents of ``fail2ban-fail2ban.cf``.

    Tunes the fail2ban daemon itself — log level, database retention — rather
    than any jail. docker-mailserver copies it to ``/etc/fail2ban/fail2ban.local``
    when it starts, so a change only applies after a restart.
    """

    content: str = ""
    restart_required: bool = True


class Fail2banConfigUpdate(BaseModel):
    """Request schema replacing the fail2ban daemon configuration."""

    content: str = Field(default="", max_length=65536)
